crates/compiler/builtins/bitcode/src/list.zig

```diff
@@ -745,37 +745,54 @@ pub fn listConcat(list_a: RocList, list_b: RocList, alignment: u32, element_widt
     } else if (list_a.isUnique()) {
         const total_length: usize = list_a.len() + list_b.len();
 
-        if (list_a.bytes) |source| {
-            const new_source = if (list_a.capacity >= total_length)
-                source
-            else
-                utils.unsafeReallocate(
-                    source,
-                    alignment,
-                    list_a.len(),
-                    total_length,
-                    element_width,
-                );
-
-            if (list_b.bytes) |source_b| {
-                @memcpy(new_source + list_a.len() * element_width, source_b, list_b.len() * element_width);
-            }
+        const resized_list_a = list_a.reallocate(alignment, total_length, element_width);
 
-            return RocList{ .bytes = new_source, .length = total_length, .capacity = total_length };
-        }
+        // These must exist, otherwise, the lists would have been empty.
+        const source_a = resized_list_a.bytes orelse unreachable;
+        const source_b = list_b.bytes orelse unreachable;
+        @memcpy(source_a + list_a.len() * element_width, source_b, list_b.len() * element_width);
+
+        // decrement list b.
+        utils.decref(source_b, list_b.len(), alignment);
+
+        return resized_list_a;
+    } else if (list_b.isUnique()) {
+        const total_length: usize = list_a.len() + list_b.len();
+
+        const resized_list_b = list_b.reallocate(alignment, total_length, element_width);
+
+        // These must exist, otherwise, the lists would have been empty.
+        const source_a = list_a.bytes orelse unreachable;
+        const source_b = resized_list_b.bytes orelse unreachable;
+
+        // This is a bit special, we need to first copy the elements of list_b to the end,
+        // then copy the elements of list_a to the beginning.
+        // This first call must use mem.copy because the slices might overlap.
```

Comment:
Is there an advantage to branching and `@memcpy` if we know they don't overlap?

Reply:
I actually think we should change all of our calls to `@memcpy` to `mem.copy`. The compiler should be smart enough to emit `@memcpy` when it is valid. If it isn't valid, `mem.copy` should still optimize to code that will run `@memcpy` if their is enough gap between the starting points of the two copy operations. Otherwise, it should optmize to `memmove`. So essentially the same cost as `@memcpy` but potentially with an extra conditional around overlap. `mem.copy` is a safer api, that won't accidentally lead to a really confusing bug when we accidentally overwrite bytes that shouldn't have been.

Reply:
have you checked if the optimized code we emit indeed does do a branch and perform memcopy/memmove if there is no overlap?

Reply:
Yeah it compiles to what looks to be the equivalent of that with the small exception that memcpy gets inlined. But otherwise, it is still fully vectored when possible.